File: app/src/test/java/org/wikipedia/analytics/eventplatform/EventPlatformClientTest.java
package org.wikipedia.analytics.eventplatform;

import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.MockedStatic;
import org.robolectric.RobolectricTestRunner;
import org.wikipedia.WikipediaApp;
import org.wikipedia.dataclient.mwapi.MwStreamConfigsResponse;
import org.wikipedia.json.GsonMarshaller;
import org.wikipedia.json.GsonUnmarshaller;
import org.wikipedia.settings.Prefs;
import org.wikipedia.test.TestFileUtil;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.greaterThanOrEqualTo;
import static org.hamcrest.Matchers.lessThanOrEqualTo;
import static org.mockito.Mockito.mockStatic;
import static org.mockito.Mockito.times;
import static org.wikipedia.analytics.eventplatform.EventPlatformClient.STREAM_CONFIGS;
import static org.wikipedia.analytics.eventplatform.EventPlatformClient.addEventMetadata;
import static org.wikipedia.analytics.eventplatform.SamplingConfig.Identifier.DEVICE;
import static org.wikipedia.analytics.eventplatform.SamplingConfig.Identifier.PAGEVIEW;
import static org.wikipedia.analytics.eventplatform.SamplingConfig.Identifier.SESSION;

@SuppressWarnings("checkstyle:magicnumber")
@RunWith(RobolectricTestRunner.class)
public class EventPlatformClientTest {

    private static final String STREAM_CONFIGS_RESPONSE = "streamconfigs_response.json";

    @Before
    public void reset() {
        STREAM_CONFIGS = new HashMap<>();
        EventPlatformClient.AssociationController.beginNewSession();

        // Set app install ID
        WikipediaApp.getInstance().getAppInstallID();
    }

    @Test
    public void testGenerateRandomId() {
        String id = EventPlatformClient.AssociationController.generateRandomId();
        assertThat(id.length(), is(20));
    }

    @Test
    public void testEventSerialization() {
        Event event = new Event("test", "test");
        addEventMetadata(event);
        String serialized = GsonMarshaller.marshal(event);
        assertThat(serialized.contains("dt"), is(true));
        assertThat(serialized.contains("app_session_id"), is(true));
        assertThat(serialized.contains("app_install_id"), is(true));
    }

    @Ignore("Disabled for testing: https://phabricator.wikimedia.org/T281001")
    @Test
    public void testOutputBufferEnqueuesEventOnSubmit() {
        Event event = new Event("test", "test");
        try (
                MockedStatic<EventPlatformClient.OutputBuffer> outputBuffer = mockStatic(EventPlatformClient.OutputBuffer.class);
                MockedStatic<EventPlatformClient.SamplingController> samplingController = mockStatic(EventPlatformClient.SamplingController.class)
        ) {
            samplingController.when(() -> EventPlatformClient.SamplingController.isInSample(event)).thenReturn(true);
            EventPlatformClient.submit(event);
            outputBuffer.verify(times(1), () -> EventPlatformClient.OutputBuffer.schedule(event));
        }
    }

    @Test
    public void testOutputBufferSendsEnqueuedEventsOnEnabled() {
        try (MockedStatic<EventPlatformClient.OutputBuffer> outputBuffer = mockStatic(EventPlatformClient.OutputBuffer.class)) {
            EventPlatformClient.setEnabled(true);
            outputBuffer.verify(times(1), EventPlatformClient.OutputBuffer::sendAllScheduled);
        }
    }

    @Test
    public void testAssociationControllerGetPageViewId() {
        String pageViewId = EventPlatformClient.AssociationController.getPageViewId();
        assertThat(pageViewId.length(), equalTo(20));
    }

    @Test
    public void testAssociationControllerGetSessionId() {
        String sessionId = EventPlatformClient.AssociationController.getSessionId();
        assertThat(sessionId.length(), equalTo(20));

        String persistentSessionId = Prefs.getEventPlatformSessionId();
        assertThat(persistentSessionId, is(notNullValue()));
        assertThat(persistentSessionId.length(), equalTo(20));
    }

    @Test
    public void testNeverInSampleIfNoStreamConfig() {
        assertThat(EventPlatformClient.SamplingController.isInSample(new Event("test", "not-configured")), is(false));
    }

    @Test
    public void testSamplingControllerGetSamplingValue() {
        double deviceVal = EventPlatformClient.SamplingController.getSamplingValue(DEVICE);
        assertThat(deviceVal, greaterThanOrEqualTo(0.0));
        assertThat(deviceVal, lessThanOrEqualTo(1.0));

        double pageViewVal = EventPlatformClient.SamplingController.getSamplingValue(PAGEVIEW);
        assertThat(pageViewVal, greaterThanOrEqualTo(0.0));
        assertThat(pageViewVal, lessThanOrEqualTo(1.0));

        double sessionVal = EventPlatformClient.SamplingController.getSamplingValue(SESSION);
        assertThat(sessionVal, greaterThanOrEqualTo(0.0));
        assertThat(sessionVal, lessThanOrEqualTo(1.0));
    }

    @Test
    public void testSamplingControllerGetSamplingId() {
        assertThat(EventPlatformClient.SamplingController.getSamplingId(DEVICE), is(notNullValue()));
        assertThat(EventPlatformClient.SamplingController.getSamplingId(PAGEVIEW), is(notNullValue()));
        assertThat(EventPlatformClient.SamplingController.getSamplingId(SESSION), is(notNullValue()));
    }

    @Ignore
    @Test
    public void testStreamConfigMapSerializationDeserialization() throws IOException {
        Map<String, StreamConfig> originalStreamConfigs = GsonUnmarshaller.unmarshal(MwStreamConfigsResponse.class,
                TestFileUtil.readRawFile(STREAM_CONFIGS_RESPONSE)).getStreamConfigs();

        Prefs.setStreamConfigs(originalStreamConfigs);
        Map<String, StreamConfig> restoredStreamConfigs = Prefs.getStreamConfigs();
        assertThat(GsonMarshaller.marshal(restoredStreamConfigs), is(GsonMarshaller.marshal(originalStreamConfigs)));
    }
}

File: app/src/test/java/org/wikipedia/analytics/eventplatform/EventPlatformClientTest.kt
package org.wikipedia.analytics.eventplatform

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.wikimedia.testkitchen.config.DestinationEventService
import org.wikimedia.testkitchen.config.StreamConfig
import org.wikimedia.testkitchen.config.StreamConfigCollection
import org.wikimedia.testkitchen.config.sampling.SampleConfig
import org.wikipedia.WikipediaApp
import org.wikipedia.analytics.eventplatform.EventPlatformClient.SamplingController
import org.wikipedia.dataclient.ServiceFactory.getAnalyticsRest
import org.wikipedia.json.JsonUtil
import org.wikipedia.settings.Prefs
import org.wikipedia.test.TestFileUtil

@RunWith(RobolectricTestRunner::class)
class EventPlatformClientTest {
    @Before
    fun reset() {
        EventPlatformClient.AssociationController.beginNewSession()

        // Set app install ID
        WikipediaApp.instance.appInstallID
    }

    @Test
    fun testGenerateRandomId() {
        val id = EventPlatformClient.AssociationController.pageViewId
        assertEquals(20, id.length)
    }

    @Test
    fun testGetStream() {
        EventPlatformClient.setStreamConfig(StreamConfig().also {
            it.streamName = "test"
        })
        assertNotNull(EventPlatformClient.getStreamConfig("test"))
        assertNull(EventPlatformClient.getStreamConfig("key.does.not.exist"))
    }

    @Test
    fun testEventSerialization() {
        val event = TestAppsEvent("test")
        val serialized = JsonUtil.encodeToString(event)!!
        assertTrue(serialized.contains("dt"))
        assertTrue(serialized.contains("app_session_id"))
        assertTrue(serialized.contains("app_install_id"))
    }

    @Test
    fun testAssociationControllerGetPageViewId() {
        val pageViewId = EventPlatformClient.AssociationController.pageViewId
        assertEquals(20, pageViewId.length)
    }

    @Test
    fun testAssociationControllerGetSessionId() {
        val sessionId = EventPlatformClient.AssociationController.sessionId
        assertEquals(20, sessionId.length)
        val persistentSessionId = Prefs.eventPlatformSessionId
        assertNotNull(persistentSessionId)
        assertEquals(20, persistentSessionId!!.length)
    }

    @Test
    fun testNeverInSampleIfNoStreamConfig() {
        assertFalse(SamplingController.isInSample(TestEvent("not-configured")))
    }

    @Test
    fun testAlwaysInSampleIfStreamConfiguredButNoSamplingConfig() {
        EventPlatformClient.setStreamConfig(StreamConfig().also {
            it.streamName = "configured"
        })
        assertTrue(SamplingController.isInSample(TestEvent("configured")))
    }

    @Test
    fun testAlwaysInSample() {
        EventPlatformClient.setStreamConfig(StreamConfig().also {
            it.streamName = "alwaysInSample"
            it.sampleConfig = SampleConfig(1.0)
        })
        assertTrue(SamplingController.isInSample(TestEvent("alwaysInSample")))
    }

    @Test
    fun testNeverInSample() {
        EventPlatformClient.setStreamConfig(
            StreamConfig().also {
                it.streamName = "neverInSample"
                it.sampleConfig = SampleConfig(0.0)
            }
        )
        assertFalse(SamplingController.isInSample(TestEvent("neverInSample")))
    }

    @Test
    fun testSamplingControllerGetSamplingValue() {
        val deviceVal = SamplingController.getSamplingValue(SampleConfig.UNIT_DEVICE)
        assertTrue(deviceVal >= 0.0)
        assertTrue(deviceVal <= 1.0)
        val pageViewVal = SamplingController.getSamplingValue(SampleConfig.UNIT_PAGEVIEW)
        assertTrue(pageViewVal >= 0.0)
        assertTrue(pageViewVal <= 1.0)
        val sessionVal = SamplingController.getSamplingValue(SampleConfig.UNIT_SESSION)
        assertTrue(sessionVal >= 0.0)
        assertTrue(sessionVal <= 1.0)
    }

    @Test
    fun testSamplingControllerGetSamplingId() {
        assertNotNull(SamplingController.getSamplingId(SampleConfig.UNIT_DEVICE))
        assertNotNull(SamplingController.getSamplingId(SampleConfig.UNIT_PAGEVIEW))
        assertNotNull(SamplingController.getSamplingId(SampleConfig.UNIT_SESSION))
    }

    @Test
    fun testGetEventService() {
        val streamConfig = StreamConfig().also {
            it.streamName = "test"
            it.destinationEventService = DestinationEventService.LOGGING
        }
        assertNotNull(getAnalyticsRest(streamConfig))
    }

    @Test
    fun testGetEventServiceDefaultDestination() {
        val streamConfig = StreamConfig().also {
            it.streamName = "test"
        }
        assertNotNull(getAnalyticsRest(streamConfig))
    }

    @Test
    fun testStreamConfigMapSerializationDeserialization() {
        val originalStreamConfigs = JsonUtil.decodeFromString<StreamConfigCollection>(TestFileUtil.readRawFile(STREAM_CONFIGS_RESPONSE))!!.streamConfigs
        Prefs.streamConfigs = originalStreamConfigs
        val restoredStreamConfigs = Prefs.streamConfigs
        assertEquals(JsonUtil.encodeToString(originalStreamConfigs), JsonUtil.encodeToString(restoredStreamConfigs))
    }

    companion object {
        private const val STREAM_CONFIGS_RESPONSE = "streamconfigs_response.json"
    }
}
